sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImplTest.java

```diff
@@ -126,6 +154,354 @@ public void testRead() throws IOException {
     verifyMetricWasSet("google.bigtable.v2.ReadRows", "ok", 1);
   }
 
+  /**
+   * This test ensures that protobuf creation and interactions with {@link BigtableDataClient} work
+   * as expected. This test checks that a single row is returned from the future.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testReadSingleRangeBelowSegmentLimit() throws Exception {
+    ByteKey start = ByteKey.copyFrom("a".getBytes(StandardCharsets.UTF_8));
+    ByteKey end = ByteKey.copyFrom("b".getBytes(StandardCharsets.UTF_8));
+
+    when(mockBigtableSource.getRanges()).thenReturn(Arrays.asList(ByteKeyRange.of(start, end)));
+    FlatRow expectedRow = FlatRow.newBuilder().withRowKey(ByteString.copyFromUtf8("a")).build();
+
+    when(mockBigtableDataClient.readFlatRows(any(ReadRowsRequest.class), any()))
+        .thenAnswer(mockReadRowsAnswer(Arrays.asList(expectedRow)));
+
+    BigtableService.Reader underTest =
+        BigtableServiceImpl.BigtableSegmentReaderImpl.create(mockSession, mockBigtableSource);
```

Comment:
If the constructor used, the mockRequest and segmentByteSize would have to be created for each test which would add another helper and some noises. Is avoiding source worth mocking the request for each test?

Reply:
Yes, this will require another helper to create the read rows request from a list of ranges. But it makes the test easier to understand...when debugging a test failure, I have one less class to worry about. Overall I think it will actually be less noise: you get to remove one mock expectation and replace it with a plain old function call that creates a request